public/app/features/alerting/unified/components/rule-editor/labels/LabelsField.tsx

```diff
@@ -0,0 +1,477 @@
+import { AlertLabels } from '../../AlertLabels';
+import { NeedHelpInfo } from '../NeedHelpInfo';
+
+import { AddButton, RemoveButton } from './LabelsButtons';
+
+const useGetOpsLabelsKeys = (skip: boolean) => {
+  const { currentData, isLoading: isloadingLabels } = labelsApi.endpoints.getLabels.useQuery(undefined, {
+    skip,
+  });
+  return { loading: isloadingLabels, labelsOpsKeys: currentData };
+};
+const useGetCustomLabels = (dataSourceName: string): { loading: boolean; labelsByKey: Record<string, Set<string>> } => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchRulerRulesIfNotFetchedYet(dataSourceName));
+  }, [dispatch, dataSourceName]);
+
+  const rulerRuleRequests = useUnifiedAlertingSelector((state) => state.rulerRules);
+  const rulerRequest = rulerRuleRequests[dataSourceName];
+
+  const labelsByKeyResult = useMemo<Record<string, Set<string>>>(() => {
+    const labelsByKey: Record<string, Set<string>> = {};
+
+    const rulerRulesConfig = rulerRequest?.result;
+    if (!rulerRulesConfig) {
+      return labelsByKey;
+    }
+
+    const allRules = Object.values(rulerRulesConfig)
+      .flatMap((groups) => groups)
+      .flatMap((group) => group.rules);
+
+    allRules.forEach((rule) => {
+      if (rule.labels) {
+        Object.entries(rule.labels).forEach(([key, value]) => {
+          if (!value) {
+            return;
+          }
+
+          const labelEntry = labelsByKey[key];
+          if (labelEntry) {
+            labelEntry.add(value);
+          } else {
+            labelsByKey[key] = new Set([value]);
+          }
+        });
+      }
+    });
+
+    return labelsByKey;
+  }, [rulerRequest]);
+
+  return { loading: rulerRequest?.loading, labelsByKey: labelsByKeyResult };
+};
+
+function mapLabelsToOptions(
+  items: Iterable<string> = [],
+  labelsInSubForm?: Array<{ key: string; value: string }>
+): Array<SelectableValue<string>> {
+  const existingKeys = new Set(labelsInSubForm ? labelsInSubForm.map((label) => label.key) : []);
+  return Array.from(items, (item) => ({ label: item, value: item, isDisabled: existingKeys.has(item) }));
+}
+
+export interface LabelsInRuleProps {
+  labels: Array<{ key: string; value: string }>;
+}
+
+export const LabelsInRule = ({ labels }: LabelsInRuleProps) => {
+  const labelsObj: Record<string, string> = labels.reduce((acc: Record<string, string>, label) => {
+    if (label.key) {
+      acc[label.key] = label.value;
+    }
+    return acc;
+  }, {});
+
+  return <AlertLabels labels={labelsObj} />;
+};
+
+export type LabelsSubformValues = {
+  labelsInSubform: Array<{ key: string; value: string }>;
+};
+
+export interface LabelsSubFormProps {
+  dataSourceName: string;
+  onClose: () => void;
+}
+
+export function LabelsSubForm({ dataSourceName, onClose }: LabelsSubFormProps) {
+  const styles = useStyles2(getStyles);
+  // get initial values from the parent form
+  const { setValue, getValues } = useFormContext<RuleFormValues>();
```

Comment:
How about passing the selected labels as an argument to the `onClose` callback instead of setting them directly using the form context?
This way we could remove the direct coupling of this component and the `RuleFormValues` form - it's already difficult to track all changes to the form fields